### ISSUE TITLE
Bug: Missing argument in route

### ISSUE THREAD

**Issue by bgeneto** [codeigniter4/CodeIgniter4#9250]
### PHP Version

8.3

### CodeIgniter4 Version

CodeIgniter 4.5.5

### CodeIgniter4 Installation Method

Composer (using `codeigniter4/appstarter`)

### Which operating systems have you tested for this bug?

Windows

### Which server did you use?

fpm-fcgi

### Database

N/A

### What happened?

The following route definition works fine as long as you don't use `url_to` or `route_to` the refer to it: 

```php
$routes->get('/test(/(:any))?', 'Test::direct/$1', ['as' => 'test']);
```
Linking as below works:
```html
<a href="/test">No parameter</a>
<a href="/test/param1">1 parameter</a>
<a href="/test/param1/param2">2 parameters</a>
```

But using both `url_to` or `route_to`  fails:

```php
<a href="<?= url_to('urlto') ?>">No parameter</a>
<a href="<?= url_to('urlto', 'param1') ?>">1 parameter</a>
<a href="<?= url_to('urlto', 'param1', 'param2') ?>">2 parameters</a>
```

### Steps to Reproduce

1. Add the new routes in `Routes.php`:
```php
$routes->get('/test(/(:any))?', 'Test::direct/$1', ['as' => 'test']);
$routes->get('/test_urlto(/(:any))?', 'Test::urlto/$1', ['as' => 'urlto']);
$routes->get('/test_routeto(/(:any))?', 'Test::routeto/$1', ['as' => 'routeto']);
```
2. Create the `Test` controller:
```php
<?php

namespace App\Controllers;

class Test extends BaseController
{
    public function direct(...$params): string
    {
        return view('test', ['params' => $params]);
    }

    public function urlto(...$params): string
    {
        return view('test_urlto', ['params' => $params]);
    }

    public function routeto(...$params): string
    {
        return view('test_routeto', ['params' => $params]);
    }
}
```

3. Create the views:
A) direct 
```html
<!DOCTYPE html>
<html lang="en">
<head>
	<meta charset="UTF-8">
	<title>Test</title>
</head>
<body>
	<h1>Testing CI4 Routing with Parameters</h1>

	<h2>Test using string directly</h2>
	<ul>
		<li>
			<a href="/test">No parameter</a>
		</li>
		<li>
			<a href="/test/param1">1 parameter</a>
		</li>
		<li>
			<a href="/test/param1/param2">2 parameters</a>
		</li>
		<li>
			<a href="/test/param1/param2/param3">3 parameters</a>
		</li>
	</ul>
	<h3>Passed Parameters</h3>
	<ul>
		<?php foreach ($params as $param) : ?>
			<li><?= $param ?></li>
		<?php endforeach; ?>
</body>
</html>
``` 
B) same view but using `url_to`

```html
<!DOCTYPE html>
<html lang="en">

<head>
	<meta charset="UTF-8">
	<title>Test</title>
</head>

<body>
	<h1>Testing CI4 Routing with Parameters</h1>
	<h2>Test using <code>url_to</code></h2>
	<ul>
		<li>
			<a href="<?= url_to('urlto') ?>">No parameter</a>
		</li>
		<li>
			<a href="<?= url_to('urlto', 'param1') ?>">1 parameter</a>
		</li>
		<li>
			<a href="<?= url_to('urlto', 'param1', 'param2') ?>">2 parameters</a>
		</li>
		<li>
			<a href="<?= url_to('urlto', 'param1', 'param2', 'param3') ?>">3 parameters</a>
		</li>
	</ul>

	<h3>Passed Parameters</h3>
	<ul>
		<?php foreach ($params as $param) : ?>
			<li><?= $param ?></li>
		<?php endforeach; ?>
</body>
</html>
```

4. Access those views at: /test and /test_urlto

### Expected Output

No `InvalidArgumentException`, `Missing argument for "(/(:any)" in route "test_urlto(/(:any))?"` while using both `url_to` and `route_to` .

### Anything else?

- One can try it with `public bool $multipleSegmentsOneParam = true;` or `public bool $multipleSegmentsOneParam = false;` in `Routing.php`. The error is the same in both situations (just the parameters are treated differently, as expected).

- One can also use a regex like `'/test(/(.+))?'` instead of `(:any)`.

**Comment by michalsn** [comment]
Using the `(:any)` placeholder in the route does not mean that the parameter value is optional. You must always put a value.

You can use: `url_to('urlto', '')`, although in general I don't think it will behave as you expect - the route will not be found.

In any case, this is not a bug.

**Comment by neznaika0** [comment]
I've been trying to use regex in routes for a long time. It failed.
Try to create several routes to get rid of the "?".
```php
$routes->get('/test/(:any)', 'Test::direct/$1', ['as' => 'test_ext']);
$routes->get('/test', 'Test::direct/$1', ['as' => 'test']);
```

To work, you need to specify a string for :any and not an array
`url_to('urlto', 'param1/param2/param3')` not `url_to('urlto', 'param1', 'param2', 'param3')`

**Comment by bgeneto** [comment]
> Using the `(:any)` placeholder in the route does not mean that the parameter value is optional. You must always put a value.
> 
> You can use: `url_to('urlto', '')`, although in general I don't think it will behave as you expect - the route will not be found.
> 
> In any case, this is not a bug.

I disagree... Because I'm using regular expressions (note the ? char). It also fails for '(/(.+))?' as mentioned in this bug report. And it works flawlessly without using url_to in links, with or without arguments. For me this is, at least, a framework inconsistency (if not a bug).

**Comment by bgeneto** [comment]
> I've been trying to use regex in routes for a long time. It failed. Try to create several routes to get rid of the "?".
> 
> ```
> $routes->get('/test/(:any)', 'Test::direct/$1', ['as' => 'test_ext']);
> $routes->get('/test', 'Test::direct/$1', ['as' => 'test']);
> ```
> 
> To work, you need to specify a string for :any and not an array `url_to('urlto', 'param1/param2/param3')` not `url_to('urlto', 'param1', 'param2', 'param3')`

Yeah! Unfortunately every time this subject comes up it is treated as a feature, read the docs etc.... 
Só I kindly suggest to remove the partial regex support from routes or explicitly says not to use the special '?' char.

**Comment by crustamet** [comment]
Well i have tested out your wanting functionality i manage to make it work with the following:

your routes are wrong in the first place and how you use the **url_to()** is wrong also.

Try it like this in routes file

	$routes->get('/test/(:any)', 'Test::direct/$1', ['as' => 'test_ext']);
	$routes->get('/test', 'Test::direct', ['as' => 'test']);

And this for the controller

	use CodeIgniter\Controller;

	class Test extends Controller
	{
		public function direct(...$params): string
		{
			echo url_to('test_ext', ...$params);
			echo '<br />';
			echo url_to('test');
			echo '<br />';
			print_r($params);die();
		}
	}
